### PR TITLE
Enrich Policy Resource

### PR DESCRIPTION
**Pull request by jonjitsu**
This PR starts adding in support for the enrich API #282 . It adds a new resource `elasticstack_elasticsearch_enrich_policy`.

### CODE REVIEW

**Comment by jonjitsu** [comment]
I've signed the contributor agreement 3 times. Once today and twice a couple weeks ago when I created the 2 character PR to fix the Makefile. I'm not sure what the next steps are to make the CLA bot happy.

**Comment by tobio** [comment]
> I'm not sure what the next steps are to make the CLA bot happy.

I can see the signature against the @jonjitsu handle. It looks like the email address associated with the commits in this PR isn't publicised in you're GH handle and so it's not linking up with the CLA sig. 

You can either rewrite the commits with the existing email associated with @jonjitsu, or include your Gmail address in your GH profile so it gets linked correctly.

File: docs/resources/elasticsearch_enrich_policy.md [hunk 1]
```diff
@@ -0,0 +1,89 @@
+---
+# generated by https://github.com/hashicorp/terraform-plugin-docs
+page_title: "elasticstack_elasticsearch_enrich_policy Resource - terraform-provider-elasticstack"
+subcategory: ""
```
Comment (tobio) [inline docs/resources/elasticsearch_enrich_policy.md]: There's a shortcoming with the tf docs generator that doesn't allow setting the subcategory from code. 

Can we add a template ([example](https://github.com/elastic/terraform-provider-elasticstack/blob/main/templates/resources/elasticsearch_index.md.tmpl#L2))for this resource setting this field.

File: internal/elasticsearch/enrich/policy.go [hunk 2]
```diff
@@ -0,0 +1,184 @@
...
+			Elem: &schema.Schema{
+				Type: schema.TypeString,
+			},
+		},
+		"query": {
+			Description:  "Query used to filter documents in the enrich index. The policy only uses documents matching this query to enrich incoming documents. Defaults to a match_all query.",
+			Type:         schema.TypeString,
+			Optional:     true,
+			ForceNew:     true,
+			ValidateFunc: validation.StringIsJSON,
```
Comment (tobio) [inline internal/elasticsearch/enrich/policy.go]:
```suggestion
			ValidateFunc: validation.StringIsJSON,
			DiffSuppressFunc: utils.DiffJsonSuppress,
```

Should we be suppressing a diff of equivalent JSON objects here?

File: internal/elasticsearch/enrich/policy.go [hunk 3]
```diff
@@ -0,0 +1,184 @@
...
+	policy, diags := elasticsearch.GetEnrichPolicy(ctx, client, compName.ResourceId)
+	if policy == nil {
+		tflog.Warn(ctx, fmt.Sprintf(`Enrich policy "%s" not found, removing from state`, compName.ResourceId))
+		d.SetId("")
+		return diags
+	}
+	if diags == nil {
+		tflog.Warn(ctx, fmt.Sprintf(`Diags nil for resource ID "%s"`, compName.ResourceId))
+		return diags
+	}
```
Comment (tobio) [inline internal/elasticsearch/enrich/policy.go]:
```suggestion
```

Why do we care about the distinction between nil/empty diags here?

**Review by tobio — COMMENTED** [review]
Looking great, thanks for taking the time to put this change together. Couple of small comments inline.

File: internal/elasticsearch/enrich/policy.go (hunk 3 above)
Comment (jonjitsu) [reply internal/elasticsearch/enrich/policy.go]: `elasticsearch.GetEnrichPolicy` can return nil when http.StatusNotFound which would make the next line `if diags.HasError()` throw some kind of null dereference error. I could make `elasticsearch.GetEnrichPolicy` return a `diag.FromErr(fmt.Error)` instead for that situation.

From looking at many other resources ex: [index.GetIndex()](https://github.com/elastic/terraform-provider-elasticstack/blob/main/internal/clients/elasticsearch/index.go#L317-L319) and it's use [resource](https://github.com/elastic/terraform-provider-elasticstack/blob/main/internal/elasticsearch/index/index.go#L916-L917) it combines the `diags != nil` test with the `model != nil` test. I could follow that pattern instead.

Comment (tobio) [reply internal/elasticsearch/enrich/policy.go]: > I could follow that pattern instead.

👍

**Pull request merged**